WordPress/WordPressTest/Services/PostServiceTests.swift

```diff
@@ -0,0 +1,153 @@
+
+        expect(posts).to(haveCount(2))
+        posts?.forEach { post in
+            expect(expectedStatuses).to(contain(post.status))
+            expect(expectedStatuses).to(contain(post.statusAfterSync))
+            expect(post.status).to(equal(post.statusAfterSync))
+        }
+    }
+
+    func testUpdatingAPostWillUpdateItsStatusAfterSyncProperty() {
+        // Arrange
+        let post = PostBuilder(context).with(statusAfterSync: .publish).drafted().withRemote().build()
+        try! context.save()
+
+        let remotePost = createRemotePost(.draft)
+        remoteMock.remotePostToReturnOnUpdatePost = remotePost
+
+        // Act
+        var postFromAPI: AbstractPost?
+        waitUntil(timeout: 3) { done in
+            self.service.uploadPost(post, success: { aPost in
+                postFromAPI = aPost
+                done()
+            }) { _ in
+                assertionFailure("This shouldn't happen.")
+            }
+        }
+
+        // Assert
+        expect(postFromAPI).notTo(beNil())
+
+        // Refetch from DB to make sure we're getting the updated data.
+        let postFromDB = context.object(with: postFromAPI!.objectID) as! AbstractPost
+        // .draft is the status because it's what the returned RemotePost has
+        expect(postFromDB.statusAfterSync).to(equal(.draft))
+        expect(postFromDB.status).to(equal(.draft))
+    }
+
+    private func createRemotePost(_ status: BasePost.Status = .draft) -> RemotePost {
+        let remotePost = RemotePost(siteID: 1,
+                                    status: status.rawValue,
+                                    title: "Tenetur im",
+                                    content: "Velit tempore rerum")!
+        remotePost.type = "qui"
+        return remotePost
+    }
+}
+
+private class PostServiceRemoteFactoryMock: PostServiceRemoteFactory {
+    var remoteToReturn: PostServiceRemote?
+
+    override func remoteForBlog(_ blog: Blog) -> PostServiceRemote? {
+        return remoteToReturn
+    }
+}
+
+private class PostServiceRemoteMock: PostServiceRemoteREST {
+    var remotePostToReturnOnGetPostWithID: RemotePost?
+    var remotePostsToReturnOnSyncPostsOfType = [RemotePost]()
+    var remotePostToReturnOnUpdatePost: RemotePost?
+
+    override func getPostWithID(_ postID: NSNumber!, success: ((RemotePost?) -> Void)!, failure: ((Error?) -> Void)!) {
+        DispatchQueue.global().async {
+            success(self.remotePostToReturnOnGetPostWithID)
+        }
+    }
+
+    override func getPostsOfType(_ postType: String!, options: [AnyHashable: Any]! = [:], success: (([RemotePost]?) -> Void)!, failure: ((Error?) -> Void)!) {
+        DispatchQueue.global().async {
+            success(self.remotePostsToReturnOnSyncPostsOfType)
+        }
+    }
+
+    override func update(_ post: RemotePost!, success: ((RemotePost?) -> Void)!, failure: ((Error?) -> Void)!) {
+        DispatchQueue.global().async {
```

Comment:
I wanted to simulate the fact that the `update()` method does not immediately call `success`. This is to avoid race conditions like:

```objective-c
        void (^successBlock)(RemotePost *post) = ^(RemotePost *post) {
            // update a state to "finished"
        }

        [remote updatePost:remotePost
                   success:successBlock
                   failure:failureBlock];
        // update the same state in `successBlock` to "in_progress"
```

If we call `successBlock` right away (in the mock), the `state` would be `"in_progress"` instead of the expected `"finished"`. 

Granted, this currently not possible with the current `PostService` code. Is it too much?

Reply:
> Granted, this currently not possible with the current PostService code. Is it too much?

Hmm, maybe? Anyway, I don't think it's something that affects the readability. I'm ok with it.